feat(scheduler): Add support for run now option

louisruch opened the pull request.

louisruch commented:
@talanknight @A-440Hz PTAL

talanknight commented on internal/scheduler/scheduler.go:
```diff
@@ -165,6 +173,19 @@ func (s *Scheduler) Start(ctx context.Context, wg *sync.WaitGroup) error {
+
 func (s *Scheduler) start(ctx context.Context) {
 	const op = "scheduler.(Scheduler).start"
 	timer := time.NewTimer(s.runJobsInterval)
```
I wonder if we should wake up the scheduler immediately when it starts instead of waiting for the full interval before running it for the first time.  Since the scheduler has just started up we don't know the last time any of the jobs were run.  Also running the jobs immediately at start up might make testing easier since our job can notify us when the scheduler has woken it up for the first time instead of having to add an arbitrary sleep in our tests, which end up being a flaky solution anyways.

louisruch replied:
Yeah this had crossed my mind as well, ill make this change

louisruch commented on internal/scheduler/scheduler.go:
```diff
@@ -177,30 +198,38 @@ func (s *Scheduler) start(ctx context.Context) {
 
-		timer.Reset(s.runJobsInterval)
+		switch {
+		case s.runNowRequested.After(start):
```
I'm not sure I follow, what I was trying to do here was not have a number of threads blocking on the `RunNow` call each writing to the `runNow` chan.  So I had the `RunNow` method attempt to write, but if nothing is receiving on that chan, the default is called and it exits. Setting the timestamp is just me making sure we honor the `RunNow` call and ensure that after calling it, the scheduler will run as soon as possible, the timer should only be set to zero if something called `RunNow` while the scheduling loop was not in the select and actually doing the scheduling. 
Happy to have a quick call to discuss if I am missing something.
As a side not the job will not run more than it is scheduled to, all this is doing is triggering the scheduling loop to look for jobs to run.

talanknight replied:
Oh you are right.  I forgot that channels are default size 0 not 1 so if there is no receiver on the other end of the channel it wont send in this case.  Instead of tracking the runNowRequested timestamps, what do you think about making the runNow channel size 1?  I get a little uneasy when we start doing timestamp comparisons across goroutines to try to handle race conditions.

louisruch replied:
Good idea, thanks

louisruch commented:
@talanknight PTAL Updating the initial timer to 0 revealed a nasty bug in the change, since I moved the schedule code out of the `select`, the `break`s where breaking out of the `for` loop instead and therefore stopping all scheduling!! 
This was happening because the repo did not have a server for a foreign key in the DB yet, I therefore also moved the start calls around in the `controller.Start`

talanknight reviewed (approved):
Good catch for the bug!  

It doesn't look like start listener depends on the jobs and scheduler system being set up so the result of the rearranging of the Start function on controller.go is that the controller can now start receiving requests before the jobs have been registered or started.  Since the scheduler is initialized prior to creating any of the repos used by the services I don't think there is a race condition.  I'm spelling this all out just for you to sanity check my understanding of the code.

louisruch commented:
> Good catch for the bug!
> 
> It doesn't look like start listener depends on the jobs and scheduler system being set up so the result of the rearranging of the Start function on controller.go is that the controller can now start receiving requests before the jobs have been registered or started. Since the scheduler is initialized prior to creating any of the repos used by the services I don't think there is a race condition. I'm spelling this all out just for you to sanity check my understanding of the code.

I updated the order to put registerJobs first, strictly speaking the scheduler should have all jobs registered before requests are received.  Start being after will have no race other than jobs wont get scheduled until it starts.
